File: tiny_agent.ts

import { Agent } from "@huggingface/tiny-agents"
import readline from "node:readline/promises"
import { stdin as input, stdout as output } from "node:process"

/*
    Create a Hugging Face agent instance.

    Model:
        Qwen2.5 72B Instruct

    Provider:
        Novita AI inference provider

    MCP Server:
        Connects to a local Gradio application exposing an MCP endpoint.

    The agent can call MCP tools exposed by that server.
*/

const agent = new Agent({
    provider: process.env.PROVIDER ?? "novita",
    model: process.env.MODEL_ID ?? "Qwen/Qwen2.5-72B-Instruct",
    apiKey: process.env.HF_TOKEN,
    servers: [{
        command: "npx",
        args: [
            "mcp-remote",
            "http://127.0.0.1:7868/gradio_api/mcp/sse",
        ],
    },],
})

const rl = readline.createInterface({ input, output });
async function main() {
    console.log("MCP has started. Type 'exit' to quit.");

    while (true) {
        const message = await rl.question("> ");

        if (["exit", "quit"].includes(message.trim().toLowerCase())) {
            break;
        }

        for await (const chunk of agent.run(message)) {
            if (typeof chunk == "string") {
                process.stdout.write(chunk);
            } else {
                process.stdout.write(JSON.stringify(chunk, null, 2));
            }
        }
        process.stdout.write("\n");
    }
    rl.close();
}
main().catch((error) => {
    console.error("Error", error);
    rl.close();
    process.exit(1);
})

File: tiny_agent.py
import asyncio
import os
from dotenv import load_dotenv

from huggingface_hub import Agent

# Load environment variables from .env file.

load_dotenv()


# Create a Hugging Face agent instance.
#
# Model:
#   Qwen2.5 72B Instruct
#
# Provider:
#   Novita AI inference provider
#
# MCP Server:
#   Connects to a local Gradio application exposing an MCP endpoint.
#
# The agent can call MCP tools exposed by that server.

agent = Agent(
    model = "Qwen/Qwen2.5-72B-Instruct", 
    provider = "novita",
    servers = [{
        "command": "npx",
        "args": [
            # MCP bridge utility
            "mcp-remote",
            
            # Local Gradio MCP endpoint
            "http://127.0.0.1:7868/gradio_api/mcp/sse"
        ]    
    }]    
)

async def main():
    """
    Simple interactive chat loop.

    Flow:
        User input
            ↓
        Agent.run()
            ↓
        LLM response stream
            ↓
        Print chunks as they arrive
    """
    
    while True:
        
        # Read user message from terminal.
        msg = input("> ")
        
        # Stream response from the agent.
        async for chunk in agent.run(msg):
            # A chunk may contain multiple choices.
            for choice in chunk.choices:
                # Print only newly generated text.
                if choice.delta.content:
                    print(f"{choice.delta.content}")
    
if __name__ == "__main__":
    # Start the asyncio event loop.
    asyncio.run(main())